### DISCUSSION TITLE
nodes/rpi-neopixels/rpi-neopixels.js

### DISCUSSION THREAD

**Discussion by kitazaki** [phoddie/node-red-mcu#75] [Q&A]
@phoddie 

config.gpio in flows.json is commented out with this commit.
https://github.com/phoddie/node-red-mcu/commit/3978cd27d8e3b8546cf055e075e229a71b0cbd15

FROM:
if (globalThis.lights)     // for compatibility with hosts that create the NeoPixel driver (M5Atom-Matrix)
            this.#np = globalThis.lights;
        else
            this.#np = new NeoPixel({length: config.pixels, pin: parseInt(config.gpio), order: config.rgb.toUpperCase()});

TO:
if (globalThis.lights) // for compatibility with hosts that create the NeoPixel driver (M5Atom-Matrix)
            this.#np = globalThis.lights;
else
            this.#np = new NeoPixel({length: config.pixels, pin: 10 /* parseInt(config.gpio) */, order: config.rgb.toUpperCase(), timing: Timing_WS2812B}); 

It previously works on Grove port of M5Stick CPlus and Core2 with NeoPixel, but not worked now because of fixed pin number 10.
I'm not sure the reason why the pin number is fixed number 10, and what about to change configurable?

![img_3499_720](https://user-images.githubusercontent.com/23477407/213921442-cb5ab250-2635-43ad-9295-366b89414262.jpg)

![img_3500](https://user-images.githubusercontent.com/23477407/213921450-4e6aa608-6405-4175-b215-5015a1ff2b73.jpg)

**Comment by phoddie** [comment]
Oops. Sorry about that. The [holiday wreath](https://github.com/phoddie/node-red-mcu/discussions/61) project uses Neopixels connected to pin 10. Because the RPi Neopixels Editor doesn't allow entering pin 10, so I change the node's source code. I shouldn't have committed that.

We're working on a Neopixels Node editor designed for Node-RED MCU Edition, so it does not have this limitation. That will avoid this problem in the future.

(Cool to see external Neopixels working on a Grove connector. I haven't tried that myself.)

**Reply by kitazaki** [reply]
that's make sense.
I've tried RPi NeoPixel node with M5Atom Matrix and M5Stack Fire already defined in device manifest.json file.
Next I've tried to chage config.gpio in flows.json to actuate NeoPixel attached on Grove port, but not worked. (M5Stick CPlus and M5Stack Core2 worked)
I agree with that great idea to chage config.gpio from NeoPixel Node editor!

**Reply by phoddie** [reply]
With today's update of both Node-RED MCU Edition and the Moddable SDK, you might take another look at this.
- There's a new MCU Neopixels node. It lets you use specify pin number for the Neopixel connection.  (It also has a more modern UI, for example to select colors and set brightness)
- Leave the Neopixels pin field empty to use the host provided Neopixels instance (e.g. globalThis.lights)
- The Moddable SDK allows overriding the host-defined Neopixel pin

I think you can only use one Neopixels strand. If the host creates globalThis.lights we may need to close it. If you run into problems, try adding that before [line 72 of the Neopixels node](https://github.com/phoddie/node-red-mcu/blob/eb06e26c65558be199703d8923770ff804176144/nodes/neopixels/mcu_neopixels.js#L72):

```js
globalThis.lights?.close();
delete globalThis.lights;
```
